Archive/Draft/RFCxxxx-PS7-Bootstrapper.md

```diff
@@ -0,0 +1,256 @@
+---
+RFC: RFCnnnn
+Author: Steve Lee
+Status: Draft
+SupercededBy: n/a
+Version: 0.1
+Area: Install
+Comments Due: Dec 31st, 2021
+Plan to implement: Yes
+---
+
+# Bootstrapper for PS7 in Windows
+
+## Motivation
+
+    As an admin,
+    I can use PS7 to provision a new Windows machine,
+    so that I can leverage new capabilities in PS7.
+
+    As a user,
+    I can use PS7 on a new Windows machine,
+    so that I can leverage new capabilities in PS7.
+
+Explicitly installing PS7 is a blocker for some customers not only due to discoverability, but also some customers have policies that
+don't allow for installing software onto Windows after initial provisioning.
+
+## Shipping in Windows
+
+There are several issues that prevent simply shipping PS7 in Windows:
+
+- Size
+- Support lifecycle differences
+
+### Size constraints
+
+A complete installation of PowerShell 7 currently is around 200 MB (unpacked).
+This is due to including everything (modules and .NET assemblies) to provide parity with what Windows PowerShell 5.1 with .NET Framework provides.
+There is currently no plan by Windows or the .NET team to ship newer .NET in Windows, so PS7 will need to continue to ship .NET with PS7.
+
+### Support lifecycle
+
+Windows has a 5+5 year support lifecycle.
+.NET supports up to 3 years for Long Term Support (LTS) versions.
+So even if size was not an issue, we cannot check in a version of PS7 with a .NET that is not supported while Windows itself is supported.
+
+### Solution
+
+To solve both of these issues, the proposal is to ship a bootstrapper in Windows that would be supported within the Windows support lifecycle.
+The boostrapper will update PS7 as needed to ensure a supported version of .NET is being used.
+Details of this bootstrapper follow.
+
+## Specification
+
+### pwsh.exe bootstrapper
+
+We would check-in a `pwsh.exe` into Windows 11+ (under `System32`) written in [native code](#native-code) that will:
+
+- Detect if PS7 is already installed
+  - If installed, pass-through all arguments to the new detached `pwsh` child process
+    - This means that if PS7 is already in the `PATH`, then that version (whether installed via MSI, MSIX, or other means)
+      needs to be updated outside of the bootstrapper
+  - Otherwise, [install](#installer) PS7 and then pass-through all arguments to the new detached `pwsh` child process
+
+The boostrapper should explicitly create a detached child `pwsh` process and then exit.
+As a detached child process, it will continue execution even after the parent has exited.
+
+Expectation is that the bootstrapper will always be first in the `$env:PATH` and executed first, so it can also serve
+to [auto-update](#auto-updating) PS7.
+
+This will allow `pwsh -c customer_script.ps1` to work seamlessly as though PS7 was available inbox.
+
+The bootstrapper would need to start the real `pwsh.exe` and bind stdin/stdout so that inputs and outputs come from `pwsh.exe` seamlessly.
```

Comment:
You would also need to ensure this works in an interactive situation where just doing `pwsh` will spawn the process that inherits the console handles itself rather than a redirected stdio.

It also means if you do something like `pwsh -c "my command"` from WinPS it will not wait until until it is finished because the parent (bootstrap) process has ended.

Reply:
good points, will make these explicit

Reply:
Another issue is how do you get the exit code to bubble itself back up to the caller. You essentially need to either have the bootstrap process continue to run until pwsh is finished and have that set it's own exit code to that of pwsh's.

Reply:
The suggestion of launching a subprocess pwsh.exe from the shim pwsh.exe has a lot of potential issues, but if we make the pwsh.exe shim a PowerShell native host process, we can actually make it behave *exactly* like the original without even spawning a subprocess. It would literally be the same as pwsh.exe but without PowerShell, and the ability to load the .NET runtime + launch PowerShell entirely from disk. I've done this before, just not with the full "let's find and download PowerShell on the fly" part.

Reply:
Yea this does sound like a better solution, I'm not too well versed in hosting the CLR side but a colleague mentioned it should be possible for the `pwsh` shim to do the bootstrapping and then just host the CLR in the same process once that step is done. You save the hit of spawning yet another process and don't need to worry about IO redirection and other things like handle inheritance and the like.

Reply:
> 

I think the shim has to stay resident and manage the pwsh lifecycle.  I have many scripts that depend on calling pwsh.exe synchronously.